src/link-preview-expander.ts

```diff
@@ -1,16 +1,142 @@
 import * as sourcegraph from 'sourcegraph'
+import { checkIsURL, cleanURL, getWord } from './util'
+import parse from 'node-html-parser'
 
-export function activate(ctx: sourcegraph.ExtensionContext): void {
-    ctx.subscriptions.add(
+export function activate(context: sourcegraph.ExtensionContext): void {
+    context.subscriptions.add(
         sourcegraph.languages.registerHoverProvider(['*'], {
-            provideHover: () => ({
-                contents: {
-                    value: 'Hello world from link-preview-expander! 🎉🎉🎉',
-                    kind: sourcegraph.MarkupKind.Markdown
+            provideHover: (document, position) => {
+                const range = document.getWordRangeAtPosition(position)
+                if (!range) {
+                    return null
                 }
-            }),
+                const maybeURL = cleanURL(getWord(document, range))
+                const isURL = checkIsURL(maybeURL)
+
+                if (!isURL) {
+                    return null
+                }
+
+                /**
+                 * Creates the markdown string to be rendered in the hover tooltip.
+                 */
+                const createResult: (metadata?: Metadata) => sourcegraph.Badged<sourcegraph.Hover> = metadata => {
+                    const { image, title, description } = metadata ?? {}
+
+                    let markdownContent = `#### [${title || maybeURL}](${maybeURL})\n\n`
+
+                    if (image || description) {
+                        markdownContent += '---\n\n'
+
+                        if (image) {
+                            markdownContent += `<img height="64" src="${image}" align="left" />`
+                        }
+
+                        if (description) {
+                            markdownContent += `<p>${description ?? ''}</p>`
```

Comment:
No HTML needed for paragraphs, just needs an empty line. Bonus is that markdown will be rendered inside it, which doesn't happen when it's inside HTML.
```suggestion
                            markdownContent += `\n\n${description ?? ''}`
```